Complete implementation of configuring IO bandwidth limits

xemul opened the pull request:
In Scylla there are two options that control IO bandwidth limit -- the /storage_service/(compaction|stream)_throughput REST API endpoints. The endpoints are partially implemented and have no counterparts in the nodetool.

This set implements the missing bits and adds tests for new functionality.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13744/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13744/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: nodetool/test_compaction_throughput
:small_blue_diamond: nodetool/test_stream_throughput
:small_blue_diamond: rest_api/test_system
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/616/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13310/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5306/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2189/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13744/)
#### Build Details:
- Duration: 4 hr 11 min
- Builder: spider3.cloudius-systems.com

denesb commented on api/compaction_manager.cc:
```diff
@@ -227,7 +219,6 @@ void unset_compaction_manager(http_context& ctx, routes& r) {
     cm::get_compaction_history.unset(r);
     cm::get_compaction_info.unset(r);
     ss::get_compaction_throughput_mb_per_sec.unset(r);
```
shouldn't `get_` be moved too?

denesb commented on test/rest_api/test_system.py:
```diff
@@ -11,3 +14,17 @@ def test_system_highest_sstable_format(rest_api):
+    resp = rest_api.send("POST", params[0], params={ params[1]: 100 })
+    resp.raise_for_status()
+    resp = rest_api.send("POST", params[0], params={ params[1]: 0 })
+    resp.raise_for_status()
```
Should also read back the value with `GET` variant to make sure it was updated.

xemul replied:
There's no get for commitlog and flush
But in general I agree, compaction and streaming have getter and can be checked, will add

denesb commented on test/nodetool/test_compaction_throughput.py:
```diff
@@ -0,0 +1,18 @@
+def test_get_compaction_throughput(nodetool, scylla_only):
+    nodetool("getcompactionthroughput", expected_requests = [
+            expected_request("GET", "/storage_service/compaction_throughput", response=0)
+        ])
```
should also check output printed by nodetool

denesb reviewed (changes requested):
I don't think we should add the API:s for memtable/commitlog stream throughput just-in-case. In general, once an API is added, we cannot remove it for compatibility, so we should have better reasons for adding them, than "might be useful".

Please make sure all the new nodetool commands have documentation in https://opensource.docs.scylladb.com/stable/operating-scylla/nodetool.html and also make sure the help of the respective commands have a reference to the online docs (see other commands for how to do it).

xemul commented:
> I don't think we should add the API:s for memtable/commitlog stream throughput just-in-case. In general, once an API is added, we cannot remove it for compatibility, so we should have better reasons for adding them, than "might be useful".

Testing. There's a test that writes lots of data into database via cql API and checks how things go. During this test commitlog code runs out of segments and stalls the seastar task queue. The root cause is still not clear, but in order to model that I need to run scylla with its outgoing IO slowed down at desired throughput and nowadays there's no way to change that

xemul commented:
upd:
- rebased
- dropped commitlog and flush throughput configuration API for future discussion
- updated nodetool docs with new commands
- added more checks in the test

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/21877)

**Changed Files:**
- [operating-scylla/nodetool-commands/getcompactionthroughput](http://previews.docs.scylladb.com/scylladb/21877/operating-scylla/nodetool-commands/getcompactionthroughput)
- [operating-scylla/nodetool-commands/getstreamthroughput](http://previews.docs.scylladb.com/scylladb/21877/operating-scylla/nodetool-commands/getstreamthroughput)
- [operating-scylla/nodetool-commands/setcompactionthroughput](http://previews.docs.scylladb.com/scylladb/21877/operating-scylla/nodetool-commands/setcompactionthroughput)
- [operating-scylla/nodetool-commands/setstreamthroughput](http://previews.docs.scylladb.com/scylladb/21877/operating-scylla/nodetool-commands/setstreamthroughput)
- [operating-scylla/nodetool](http://previews.docs.scylladb.com/scylladb/21877/operating-scylla/nodetool)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

xemul commented:
upd:
- fixed title overline in new docs

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13811/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13811/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: nodetool/test_compaction_throughput
:small_blue_diamond: nodetool/test_stream_throughput
:small_blue_diamond: rest_api/test_system
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/80/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2281/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/678/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13410/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5399/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13811/)
#### Failed Tests (1/39812):
- [test_index_doesnt_flood_cache_in_small_partition_workload](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13811/testReport/junit/boost/cache_algorithm_test/test_index_doesnt_flood_cache_in_small_partition_workload) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_index_doesnt_flood_cache_in_small_partition_workload)
#### Build Details:
- Duration: 4 hr 5 min
- Builder: spider7.cloudius-systems.com

xemul commented:
#21536

regevran commented:
@denesb:
> I don't think we should add the API:s for memtable/commitlog stream throughput just-in-case.

@xemul:
> in order to model that I need to run scylla with its outgoing IO slowed down at desired throughput

Do we have any other means to change the throughput?
The problem this PR is supporting is a long standing one and I do think we need tools for investigation.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13851/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13851/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: nodetool/test_compaction_throughput
:small_blue_diamond: nodetool/test_stream_throughput
:small_blue_diamond: rest_api/test_system
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/709/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13441/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2312/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5431/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/111/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13851/)
#### Build Details:
- Duration: 4 hr 9 min
- Builder: spider1.cloudius-systems.com

denesb commented:
> > I don't think we should add the API:s for memtable/commitlog stream throughput just-in-case. In general, once an API is added, we cannot remove it for compatibility, so we should have better reasons for adding them, than "might be useful".
> 
> Testing. There's a test that writes lots of data into database via cql API and checks how things go. During this test commitlog code runs out of segments and stalls the seastar task queue. The root cause is still not clear, but in order to model that I need to run scylla with its outgoing IO slowed down at desired throughput and nowadays there's no way to change that

Ok, if there is an existing need, I suppose we can add it.

xemul commented:
@annastuchlik please review the docs part.

xemul commented:
@scylladb/scylla-maint , please consider merging
